feat: add semgrep code scanning via -safe argument

ericrallen opened the pull request:
### Describe the changes you have made:

This PR introduces the first tool in a safety toolkit to help verify that code _might_ be safe to execute.

It adds a `-safe` argument that can be used to enable code scanning via [semgrep](https://semgrep.dev/).

`-safe` has 3 possible values:

- `off`: (_default_) don't enabled `safe_mode`
- `auto`: enable `safe_mode` and always scan code with `semgrep` before asking to execute
- `ask`: enable `safe_mode` and ask the user if they want to scan a code snippet with `semgrep` before asking to execute


**Note**: The `-safe` option is disabled by default and is entirely opt-in. Enabling `safe_mode` disables `auto_run`.

### Reference any relevant issue (Replaces #24)

- [x] I have performed a self-review of my code:

### I have tested the code on the following OS:
- [ ] Windows
- [x] MacOS
- [ ] Linux

### AI Language Model (if applicable)
- [x] GPT4
- [x] GPT3
- [ ] Llama 7B
- [ ] Llama 13B
- [ ] Llama 34B
- [ ] Huggingface model (Please specify which one)

I tested this with code generated and read from the filesystem with `gpt-3.5-turbo` and `gpt-4`, but the code scanning doesn't include the model in any way, so it should function regardless of what model the user has configured.

ericrallen commented on interpreter/cli/cli.py:
```diff
@@ -84,7 +92,10 @@ def cli(interpreter):
             parser.add_argument(f'-{arg["nickname"]}', f'--{arg["name"]}', dest=arg["name"], help=arg["help_text"], action='store_true')
         else:
-            parser.add_argument(f'-{arg["nickname"]}', f'--{arg["name"]}', dest=arg["name"], help=arg["help_text"], type=arg["type"])
+            choices = arg["choices"] if "choices" in arg else None
+            default = arg["default"] if "default" in arg else None
+
+            parser.add_argument(f'-{arg["nickname"]}', f'--{arg["name"]}', dest=arg["name"], help=arg["help_text"], type=arg["type"], choices=choices, default=default)
```
There wasn't a way to add choices and defaults with the current code, but I didn't want to add more conditional branches, so this is what I came up with. Happy to adjust this to make it more Pythonic if there's a better approach.

ericrallen commented on interpreter/code_interpreters/language_map.py:
```diff
@@ -0,0 +1,17 @@
+from .languages.r import R
+
+
+language_map = {
```
I ended up needing to reference the same language map that the `create_code_interpreter` module was using, so I moved it into it's own module to make it easier to share.

ericrallen commented:
![Screen Shot 2023-09-23 at 12 33 40 AM](https://github.com/KillianLucas/open-interpreter/assets/1667415/0e88467b-0875-4829-a8a1-687fde812b4b)

ericrallen commented on interpreter/cli/cli.py:
```diff
@@ -124,6 +135,10 @@ def cli(interpreter):
         if attr_value is not None and hasattr(interpreter, attr_name):
             setattr(interpreter, attr_name, attr_value)
 
+    # if auto_run is enabled, we won't bother scanning code
+    if interpreter.auto_run and not interpreter.scan_code == "off":
+        setattr(interpreter, "scan_code", "off")
```
Not sure if there's a better pattern to handle interconnected attributes like this, but I'm guessing there might be a few things in the future that might depend on the status of `auto_run`, so we might want to explore a less naive approach than what I've done here.

ericrallen commented on interpreter/core/core.py:
```diff
@@ -30,6 +30,7 @@ def __init__(self):
         self.auto_run = False
         self.debug_mode = False
         self.max_output = 2000
+        self.scan_code = "off"
```
I wasn't sure if we'd want to put this under some sort of larger `security`/`safety` subsection so that the `config.yaml` is a little easier to understand and doesn't become too polluted with top-level keys as more features (like using `guarddog` to scan dependencies before installing them) are added.

Maybe something like:

```yaml
system_message: |
  You are Open Interpreter, a world-class programmer that can complete any goal by executing code.
  ...
local: false
model: "gpt-4"
temperature: 0
security:
  scan_code: "ask"
```

But I didn't want to start fiddling with the config parsing logic if I didn't need to.

ericrallen commented on interpreter/code_interpreters/languages/python.py:
```diff
@@ -4,6 +4,8 @@
 import re
 
 class Python(SubprocessCodeInterpreter):
+    file_extension = "py"
```
These `file_extension`s aren't strictly necessary, but it felt nicer when generating the temporary file for semgrep to scan to do it with the appropriate extension and could come in handy for some future ideas - like adding some functionality to directly export generated code into a file via a some `%magic` command without needing to have another exchange with Open Interpreter to get it to create some code to pipe the other code into the file.

In the case of something like the JavaScript language configuration this could get a little more complicated since they could also be modules that use the `mjs` extension, but for the purposes it's being used for right now - a naive extension suffices.

ericrallen commented:
**NOTE**: This depends on #511 to work as expected.

KillianLucas commented:
@ericrallen this is brilliantly, beautifully done. To me this is a strong foundation for a `--safe` flag. 

I'm in favor of fewer, simpler options so I'm going to remove the tiers. `--safe` will simply scan the code automatically (let me know if this is very important for certain workflows, or if `semgrep` takes a very long time which would justify the tiers). We can add additional security features into it later.

Just to confirm before merging this (and #511), users don't have to log in to semgrep to use it?

ericrallen commented:
@KillianLucas I've cleaned this one up a bit given the desire to put safety tools under a single flag in the future - I've also simplified the output and made `safe_mode` take precedence over `auto_run` given that automatically running code is the opposite of safe.

I moved the temporary file creation and cleanup into it's own utility since there are other safety tools, like scanning for exposed secrets, that will likely need to leverage a similar temporary file functionality.

Semgrep's [free tier](https://semgrep.dev/pricing/) works really well for this usecase - just scanning a snippet of code locally - but the CLI also allows users who have a Semgrep account to login and leverage some of the more advanced features if they want.

Here's some recordings of the output for each value of the flag:

### Open Interpreter with `-safe "off"`:
![OpenInterpreter-SafeMode-Off](https://github.com/KillianLucas/open-interpreter/assets/1667415/daa01e0f-f804-40eb-aae5-e41c05493bb4)

### Open Interpreter with `-safe "ask"`:
![OpenInterpreter-SafeMode-Ask](https://github.com/KillianLucas/open-interpreter/assets/1667415/5bf7d1b5-31fa-4082-bd9f-c98438e94b82)

### Open Interpreter with `-safe "auto"`:
![OpenInterpreter-SafeMode-Auto](https://github.com/KillianLucas/open-interpreter/assets/1667415/b56eb4c7-cfc6-453e-aff4-4de7bade6e13)

ericrallen commented:
Here are the prompts I've been using to quickly test this functionality:

## Generate Safe (hopefully) Code

```
Solve FizzBuzz for the numbers 0 through 17 with a basic Python loop and if statement. No need to explain the rules of the game, just show us the code.
```

## Generate Vulnerable Code

```
We just added a code scanner to Open Interpreter that relies on Semgrep to detect vulnerable code. Please generate a Python script with an obvious vulnerability, like SQL injection for example, that Semgrep should be able to identify. DO NOT MENTION THE VULNERABILITY IN THE SCRIPT VARIABLE NAMES OR CODE COMMENTS. This script should look like a developer accidentally included the vulnerability without realizing it. DO NOT WARN US ABOUT THE VULNERABILITY OR ITS IMPLICATIONS. We want to see if the code scanner identifies the vulnerability without you alerting us to it.
```

ericrallen commented:
You can also save the following snippets to files on your machine and ask Open Interpreter to read and execute the code.

**Note**: These examples were generated by Open Interpreter as examples of vulnerable code that should be identified by the code scanner.

## JavaScript

```js
const vm = require('vm');
const contextObject = { globalVar: 1 };

// safe
vm.runInContext('globalVar *= 2;', contextObject);

// vulnerable
let userInput = 'this.constructor.constructor("return process.env")()'; // Value supplied by user input
vm.runInContext(`globalVar = ${userInput};`, contextObject);

// safe
const code = `return 'hello ' + name`
vm.compileFunction(code, [], { parsingContext: vm.createContext({ name: 'name' }) })

// vulnerable
let userInput = '1; while (true)
```

## Python

```py
import subprocess
import sys

# Vulnerable
user_input = "foo && cat /etc/passwd" # value supplied by user
subprocess.call("grep -R {} .".format(user_input), shell=True)

# Vulnerable
user_input = "cat /etc/passwd" # value supplied by user
subprocess.run(["bash", "-c", user_input], shell=True)

# Not vulnerable
user_input = "cat /etc/passwd" # value supplied by user
subprocess.Popen(['ls', '-l', user_input])

# Not vulnerable
subprocess.check_output('ls -l dir/')
```

ericrallen commented:
@KillianLucas I rebased this branch, fixed a bit of code that I forgot to update from `scan_code` to `safe_mode`, and added a loading indicator during the scan so that the user has some feedback that things are happening. This used to display the `semgrep` output, but that had a bit too much noise and it felt like it needed something to let the user know that things are happening.

## Loading Indicator

![OpenInterpreter-SafeMode-Auto-Loading](https://github.com/KillianLucas/open-interpreter/assets/1667415/989149f5-737a-4436-b62e-7a90cb6afb9a)

Also here's a simple prompt for anyone who wants to test this to get right into a state where you can test the scanner after running with `poetry run interpreter -safe "ask"` or `poetry run interpreter -safe "auto"`:

> Solve FizzBuzz for 0 through 17. Don't explain the code or tell me your process or how FizzBuzz works. Just generate the code so we can execute it.

ericrallen commented on interpreter/code_interpreters/languages/python.py:
```diff
@@ -4,6 +4,9 @@
 
 class Python(SubprocessCodeInterpreter):
+    file_extension = "py"
+    proper_name = "Python"
```
These proper names are mostly just being used for the code scanner to say

> Code Scanner: No Issues were found with this Python code

We could just rely on the lowercase language string from the code block, but it felt nicer and more human-readable this way.

ericrallen commented on interpreter/terminal_interface/terminal_interface.py:
```diff
@@ -8,13 +8,25 @@
+            interpreter_intro_message.append(f"**Safe Mode**: {interpreter.safe_mode}")
+        else:
+            interpreter_intro_message.append(
+                "Use `interpreter -y` or set `auto_run: true` to bypass this."
```
Since `safe_mode` is opt-in and disables `auto_run` it seemed unnecessary to mention the `-y` flag.

ericrallen commented:
Not sure why the workflow tests are failing. Everything seems to work when I run the test suite locally.

ericrallen commented:
@KillianLucas I’ll rebase again tomorrow to resolve the `poetry.lock`conflict, and try to see if I can get to the bottom of what’s wrong with the test suite, but this PR only applies features behind a flag, so I’m not sure how it impacted them.

Any guidance on getting it ready to merge would be appreciated.

KillianLucas commented:
GREAT WORK ERIC! Literally our most impressive and extensive PR on this whole project to date. Extreme utility here, and you've opened the door to a substantial wing of the OI project (and maybe even to an industry/discipline??) focused on making LLM-written code **safe**. Merging now. 🎉
